prevent python view impls from getting registered to the meta key

bdhirsh opened the pull request:
We don't want to register view ops in python to the `Meta` dispatch key, because doing that prevents us from correctly aliasing storage information. This PR fixes the existing python registrations, and makes it an error to do that in the future. Example:
```
with FakeTensorMode.push() as mode:
    b = torch.ones(2)
    c = b.unsqueeze(-1)
    b_ = StorageWeakRef(b.storage())
    c_ = StorageWeakRef(c.storage())
    print(b_.cdata)
    print(c_.cdata)  # their storages are different (now fixed in this PR)
```


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #80897
* #82009
* #82008
* __->__ #82007

bdhirsh committed:
prevent python view impls from getting registered to the meta key

[ghstack-poisoned]

eellison commented on torch/_decomp/__init__.py:
```diff
@@ -116,6 +116,15 @@ def add_op_to_table(aten_op):
+                    if any(a.alias_info is not None and not a.alias_info.is_write for a in op_overload._schema.arguments):
+                        raise RuntimeError(f"""
+Attempting to register a python meta kernel for a view operator: {str(op_overload)}.
+We shouldn't do this, because the output will report as not having aliased storages.
```
I would maybe add a note that this is the state of the current implementation but could be fixed in the future.

eellison reviewed (approved):
LGTM one comment

eellison commented:
fix lint and land ?

bdhirsh committed:
Update on "prevent python view impls from getting registered to the meta key"

We don't want to register view ops in python to the `Meta` dispatch key, because doing that prevents us from correctly aliasing storage information. This PR fixes the existing python registrations, and makes it an error to do that in the future. Example:
```
with FakeTensorMode.push() as mode:
    b = torch.ones(2)
    c = b.unsqueeze(-1)
    b_ = StorageWeakRef(b.storage())
    c_ = StorageWeakRef(c.storage())
    print(b_.cdata)
    print(c_.cdata)  # their storages are different (now fixed in this PR)
```





[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
